src/main.c

```diff
@@ -879,6 +1094,74 @@ void handleSign(uint8_t p1, uint8_t p2, uint8_t *workBuffer,
     *flags |= IO_ASYNCH_REPLY;
 }
 
+void handleSignMessage(uint8_t p1, uint8_t p2, uint8_t *workBuffer,
+                uint16_t dataLength, volatile unsigned int *flags,
+                volatile unsigned int *tx) {
+    UNUSED(tx);
+    bool last = (p1 & P1_LAST);
+    uint32_t i;
+    p1 &= 0x7F;
+    if (p1 == P1_FIRST) {
+        tmpCtx.messageContext.pathLength = workBuffer[0];
+        if ((tmpCtx.messageContext.pathLength < 0x01) ||
+            (tmpCtx.messageContext.pathLength > MAX_BIP32_PATH)) {
+            PRINTF("Invalid path\n");
+            THROW(0x6a85);
+        }
+        workBuffer++;
+        dataLength--;
+        for (i = 0; i < tmpCtx.messageContext.pathLength; i++) {
+            tmpCtx.messageContext.bip32Path[i] =
+                (workBuffer[0] << 24) | (workBuffer[1] << 16) |
+                (workBuffer[2] << 8) | (workBuffer[3]);
+            workBuffer += 4;
+            dataLength -= 4;
+        }
+        if (((p2 & P2_SECP256K1) == 0) && ((p2 & P2_ED25519) == 0)) {
+            THROW(0x6B01);
+        }
+        if (((p2 & P2_SECP256K1) != 0) && ((p2 & P2_ED25519) != 0)) {
+            THROW(0x6B02);
+        }
+        tmpCtx.messageContext.curve =
+            (((p2 & P2_ED25519) != 0) ? CX_CURVE_Ed25519 : CX_CURVE_256K1);
+    } else 
+    if (p1 != P1_MORE) {
+        THROW(0x6B03);
+    }
+
+    if (p1 == P1_FIRST) {
+        tmpCtx.messageContext.rawMessageLength = dataLength;
+        os_memmove(tmpCtx.messageContext.rawMessage, workBuffer, dataLength);
+    }
+    else
+    if (p1 == P1_MORE) {
+        if ((tmpCtx.messageContext.rawMessageLength + dataLength) > MAX_RAW_MESSAGE) {
+            THROW(0x6A81);
```

Comment:
instead start sha256 computation on the fly, and free tmpCtx.messageContext

Reply:
Not relevant anymore, see https://github.com/ArkEcosystem/ledger/pull/4#discussion_r273913305